SpherePacking/MagicFunction/b/psi.lean

```diff
@@ -51,71 +51,406 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+    have resolve' : -(h3 * d) + -(h2 * d) = - ((h2 + h3) * d) := by
+      rw [← @neg_add, @right_distrib, add_comm]
+    rw [resolve', mul_neg, ← neg_div', pow_two, ← mul_assoc 128 _ _, mul_comm h4 d,
+      mul_assoc d h4 (d * h4), ← div_div, ← mul_div_assoc', div_self, mul_one,
+      mul_comm d h4, ← mul_assoc h4 h4 d, ← div_div, mul_div_assoc, ← pow_two, neg_div']
+    exact dnonzero
+  rw [resolve]
+  · have resolve' (h2 h3 h4 d : ℂ) (dnonzero : d ≠ 0) :
+      128 * ( - (h2 * d) + (h4 * d) ) / (h3 * d) ^ 2 = (128 * ((h4 - h2) / (h3) ^ 2 )) / d := by
+      have resolve'' : - (h2 * d) + (h4 * d) = (h4 - h2) * d := by
+        rw [@neg_add_eq_sub, ← sub_mul]
+      rw [resolve'', ← mul_assoc 128 _ _, pow_two, mul_comm h3 d, mul_assoc d h3 (d * h3),
+        ← div_div, ← mul_div_assoc', div_self, mul_one, mul_comm d h3,
+        ← mul_assoc h3 h3 d, ← div_div, mul_div_assoc, ← pow_two ]
+      exact dnonzero
+    rw [resolve']
+    · rw [← add_div, div_mul, div_self]
+      · rw [div_one, ← mul_neg, ← mul_add, add_comm]
+        simp only [mul_eq_mul_left_iff, OfNat.ofNat_ne_zero, or_false]
+        rw [@Mathlib.Tactic.RingNF.add_neg]
+      · exact pow_ne_zero 2 (UpperHalfPlane.ne_zero z)
+    · exact pow_ne_zero 2 (UpperHalfPlane.ne_zero z)
+  · exact pow_ne_zero 2 (UpperHalfPlane.ne_zero z)
+-- proof of ψS_eq' complete.
+
+lemma ψS_eq : ψS = 128 * (- ((H₂_MF + H₃_MF) / H₄_MF ^ 2) - (H₂_MF - H₄_MF) / (H₃_MF ^ 2)) := by
+  rw [ψS_eq', sub_eq_add_neg (H₄_MF : ℍ → ℂ), add_comm (H₄_MF : ℍ → ℂ) _,
+    ← sub_neg_eq_add, ← neg_sub', neg_div, ← neg_add', add_comm, neg_add']
+-- proof of ψS_eq complete.
 
 end eq
 
--- TODO: Define all the slash relations between the `ψ` functions.
-
 section rels
 
-lemma ψT_slash_T : ψT ∣[-2] T = ψI := by sorry
-lemma ψS_slash_S : ψS ∣[-2] S = ψI := by sorry
-lemma ψS_slash_ST : ψS ∣[-2] (S * T) = ψT := by sorry
-lemma ψS_slash_T : ψS ∣[-2] T = ψS := by sorry
-lemma ψT_slash_S : ψT ∣[-2] S = -ψT := by sorry
-lemma ψI_slash_TS : ψI ∣[-2] (T * S) = -ψT := by sorry
-lemma ψS_slash_STS : ψS ∣[-2] (S * T * S) = -ψT := by sorry
-lemma ψS_slash_TSTS : ψS ∣[-2] (T * S * T * S) = -ψT := by sorry
+lemma ψT_slash_T : ψT ∣[-2] T = ψI := by
+  ext z
+  rw [ψT_eq , ψI_eq , slashT']
+  simp only [Pi.mul_apply, Pi.ofNat_apply, Pi.add_apply, Pi.div_apply,
+    Pi.pow_apply, Pi.smul_apply, Pi.sub_apply, smul_add, nsmul_eq_mul, Nat.cast_ofNat]
+  rw [← slashT z ⇑H₂_MF, ← slashT z ⇑H₃_MF, ← slashT z ⇑H₄_MF]
+  have hh2 : (H₂_MF : ℍ → ℂ) = H₂ := by exact rfl
+  have hh3 : (H₃_MF : ℍ → ℂ) = H₃ := by exact rfl
+  have hh4 : (H₄_MF : ℍ → ℂ) = H₄ := by exact rfl
+  rw [hh2, hh3, hh4, H₂_T_action, H₃_T_action, H₄_T_action]
+  field_simp -- if I replace this by the result of simp?, the proof no longer works.
+  rw [← mul_div_assoc', ← mul_div_assoc', ← mul_add, add_comm (H₄ z) (H₃ z),
+    add_comm  (- (H₂ z)) (H₄ z), Mathlib.Tactic.RingNF.add_neg]
```

Comment:
You can merge these two lines either by incorporating things in the second line into the first, as I have done below, or by doing `field_simp` followed by `ring_nf`. Both of these would be acceptable.

A small remark: lemmas like `Mathlib.Tactic.RingNF.add_neg` should not be referenced explicitly---not here or anywhere else. These are lemmas that are specifically meant to be used by the `ring_nf` tactic. In fact, if you go to its proof, you will find that it is really just `sub_eq_add_neg`.

```suggestion
  field_simp [← mul_div_assoc', ← mul_add, add_comm (H₄ z) (H₃ z), add_comm  (- (H₂ z)) (H₄ z),
    sub_eq_add_neg]
```